PER-8720: FusionAuth Index Page

meisekimiu opened the pull request:
<del>why did I make this a separate branch off of main when I have the other branch's code still on the dev instance. why would I do this</del>

This is the re-themed version of the FusionAuth index page for Permanent's instance. The page includes a link to the main web-app as well as the lock icon link to the admin section.

The original FusionAuth index template features a rather notable inclusion in its footer element, which I have mostly kept preserved in this restyling. So I'd like to open up what is probably the most silly and pointless PR discussion ever: <span title="Personally I would never put an easter egg in the Permanent app myself, but also I think easter eggs are precious and I don't want to just destroy one, you know? ...... wait... does this hover text count as an easter egg????">**Should we keep in FusionAuth's default easter egg or remove it?** I've left it in for this initial commit because I don't want to single-handedly make a "no fun allowed" judgement call here, but I feel like the easter egg is even *more* confusing in the context of our page and audience in particular. BUT... normal users should never even reach this page anyway without URL hacking.</span>

Resolves PER-8720. Interestingly enough, the ticket also mentions the "change password complete" page, which is already done. I assume that's because of the weirdness of some of the FusionAuth theming tickets. I think?

You can test this by just going to the Permanent FusionAuth Dev Instance's index page!

meisekimiu committed:
FusionAuth Index Page

This is the re-themed version of the FusionAuth index page for 
Permanent's instance.
The original index page's template features an... interesting addition 
in the footer, which is still included in this version of the template.

jasonaowen commented on theme/index.ftl:
```diff
@@ -0,0 +1,23 @@
+    [#-- Using an HTML comment so this is visible in the rendered HTML. We don't want to freak anyone out. --]
+    <!-- Feel free to remove this, it is just a stupid easter egg. Enjoy. -->
+    <!--   "Escape is impossible when you're caught in the net" https://www.imdb.com/title/tt0113957/  -->
+    <div style="position:fixed; left:0; bottom: 0; margin-bottom: 7px; margin-left: 10px;"><a style="color: #fff;" href="https://the-praetorians.net" title="Escape is impossible when you're caught in the net.">π</a></div>
```
oh my gosh this is amazing!

slifty replied:
Ha!!!

So, as a person who very much values fun I will mix it up and play the potential wet blanket question role: would this element be visible to folks with screen readers?  If so I think we should ditch it.

jasonaowen replied:
So there is the [ARIA `presentation="none"`](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/presentation_role) attribute, but since this is a link that won't work:

> The [`<a>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a) is a special case. Its role would have been exposed even if it had the `presentation` or `none` role directly applied to it. Browsers ignore `role="presentation"` and `role="none"` on focusable elements, including link and inputs, or anything with a [tabindex](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/en-US/docs/Web/HTML/Global_attributes/tabindex) attribute set.Browsers also ignore the inclusion of the role if any of the element contains any global ARIA states and properties, such as [`aria-describedby`](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-describedby).

There's the [`aria-hidden` attribute](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-hidden), but that similarly warns:

> Do not use `aria-hidden="true"` on focusable elements.

More generally, they write:

> You should not be hiding visible content unless doing so improves the experience for users of assistive technologies by removing redundant or extraneous content. Only when identical or equivalent meaning and functionality is exposed to assistive technologies can removing visible content from the accessibility API be considered.

I don't know how this applies to easter eggs; the content is rendered but is intentionally zero-contrast to make it hard to discover.

Seems a shame to lose this bit of fun, though. Maybe we only leave it on the dev site, and not on prod? And therefore not check it in here? I don't know.

jasonaowen commented:
> Resolves PER-8720. Interestingly enough, the ticket also mentions the "change password complete" page, which is already done. I assume that's because of the weirdness of some of the FusionAuth theming tickets. I think?

I think you're talking about [PER-8537](https://permanent.atlassian.net/browse/PER-8537), and I definitely miscommunicated around that in a way that we never went back and fixed; I was using that issue as a template to clone for other theming tickets, and I intended for the root to be themed in [PER-8721](https://permanent.atlassian.net/browse/PER-8721). Instead, 8721 got closed as a duplicate. So yes, there's some noise in 8537, because I didn't intend for it to be worked on directly!

jasonaowen commented on theme/index.ftl:
```diff
@@ -0,0 +1,23 @@
+<html>
+  [@helpers.permHeadTag title="Permanent.org - FusionAuth"/]
+  <body class="permanent">
+    [@helpers.permHeader showAdminLock=true/]
```
Ah, `showAdminLock` is an elegant way to reuse this code - nice!

jasonaowen commented on theme/index.ftl:
```diff
@@ -0,0 +1,23 @@
+    <main style="text-align: center;">
+      <h1>Welcome!</h1>
+      <p>The Permanent web application can be found on <a href="https://permanent.org/app">Permanent.org</a></p>
+      [#-- Maybe put a link to account self service here? --]
```
Yes, great idea - once we get that all set up!

jasonaowen commented:
This is live at the dev site: https://permanent-dev.fusionauth.io/ . And it looks good to me!

![Screen Shot 2022-03-25 at 14 51 12](https://user-images.githubusercontent.com/1494855/160207232-5f96e0b9-cf7d-46b3-a8e0-ffea6f512433.png)

jasonaowen commented on theme/helpers.ftl:
```diff
@@ -1062,13 +1062,18 @@
         </a>
+        [#if showAdminLock]
+          <a href="/admin/" title="Administrative Login">
+            <i class="fa fa-lock" style="font-size: 18px; color: #fff; opacity: 0.5;"></i>
```
Is `fa-lock` a CSS class provided by FusionAuth?

Can we give it a bit more padding on the right? It looks a bit too close to the edge when the horizontal resolution is below 850px or so.

meisekimiu replied:
`fa` and `fa-lock` is a Font Awesome  class I believe! It is weirdly coincidental that both have the same initials now that I think about it.

I will add more padding to this though!

jasonaowen replied:
Oh ha, that makes sense!

Thank you!

meisekimiu committed:
Show default FusionAuth easter egg only in dev

This will only show the funny FusionAuth easter egg on our dev instance, 
which means no usability concerns or user confusion in production, and 
we still get to keep a bit of joy in our lives!

meisekimiu commented:
It seems like we never really reached a conclusion to the "Keep the Easter Egg or Not?" debate in this PR. I did like Jason's suggestion to only keep it in dev, but I wasn't sure about not checking that in. So, I posted a commit here to only render the easter egg in our dev FusionAuth Admin instance, meaning we can keep the easter egg in some form, but it won't show up in production at all!

Edit: I forgot to mention, it's also up on dev too!

jasonaowen commented on theme/index.ftl:
```diff
@@ -15,9 +15,11 @@
-    <!-- Feel free to remove this, it is just a stupid easter egg. Enjoy. -->
-    <!--   "Escape is impossible when you're caught in the net" https://www.imdb.com/title/tt0113957/  -->
-    <div style="position:fixed; left:0; bottom: 0; margin-bottom: 7px; margin-left: 10px;"><a style="color: #fff;" href="https://the-praetorians.net" title="Escape is impossible when you're caught in the net.">π</a></div>
+    [#if tenantId == "0d4161cf-c520-e5fa-9250-e23ba0e3e6a3" ]
```
This is an elegant solution!

jasonaowen reviewed (approved):
This looks great to me! Thanks for raising a potential problem and then finding a good middle ground, @meisekimiu!
